File: BeanCovertor.swift
//
//  BeanCovertor.swift
//  RCTAgora
//
//  Created by LXH on 2020/4/9.
//  Copyright © 2020 Syan. All rights reserved.
//

import AgoraRtcKit
import Foundation

func mapToPoint(_ map: [String: Any]) -> CGPoint {
    var point = CGPoint()
    if let x = map["x"] as? NSNumber {
        point.x = CGFloat(truncating: x)
    }
    if let y = map["y"] as? NSNumber {
        point.y = CGFloat(truncating: y)
    }
    return point
}

func mapToSize(_ map: [String: Any]) -> CGSize {
    var size = CGSize()
    if let width = map["width"] as? NSNumber {
        size.width = CGFloat(truncating: width)
    }
    if let height = map["height"] as? NSNumber {
        size.height = CGFloat(truncating: height)
    }
    return size
}

func mapToRect(_ map: [String: Any]) -> CGRect {
    return CGRect(
        origin: mapToPoint(map),
        size: mapToSize(map)
    )
}

func mapToVideoEncoderConfiguration(_ map: [String: Any]) -> AgoraVideoEncoderConfiguration {
    let config = AgoraVideoEncoderConfiguration()
    if let dimensions = map["dimensions"] as? [String: Any] {
        config.dimensions = mapToSize(dimensions)
    }
    if let frameRate = map["frameRate"] as? NSNumber {
        config.frameRate = frameRate.intValue
    }
    if let minFrameRate = map["minFrameRate"] as? NSNumber {
        config.minFrameRate = minFrameRate.intValue
    }
    if let bitrate = map["bitrate"] as? NSNumber {
        config.bitrate = bitrate.intValue
    }
    if let minBitrate = map["minBitrate"] as? NSNumber {
        config.minBitrate = minBitrate.intValue
    }
    if let orientationMode = map["orientationMode"] as? NSNumber {
        if let orientationMode = AgoraVideoOutputOrientationMode(rawValue: orientationMode.intValue) {
            config.orientationMode = orientationMode
        }
    }
    if let degradationPreference = map["degradationPrefer"] as? NSNumber {
        if let degradationPreference = AgoraDegradationPreference(rawValue: degradationPreference.intValue) {
            config.degradationPreference = degradationPreference
        }
    }
    if let mirrorMode = map["mirrorMode"] as? NSNumber {
        if let mirrorMode = AgoraVideoMirrorMode(rawValue: mirrorMode.uintValue) {
            config.mirrorMode = mirrorMode
        }
    }
    return config
}

func mapToBeautyOptions(_ map: [String: Any]) -> AgoraBeautyOptions {
    let options = AgoraBeautyOptions()
    if let lighteningContrastLevel = map["lighteningContrastLevel"] as? NSNumber {
        if let lighteningContrastLevel = AgoraLighteningContrastLevel(rawValue: lighteningContrastLevel.uintValue) {
            options.lighteningContrastLevel = lighteningContrastLevel
        }
    }
    if let lighteningLevel = map["lighteningLevel"] as? NSNumber {
        options.lighteningLevel = lighteningLevel.floatValue
    }
    if let smoothnessLevel = map["smoothnessLevel"] as? NSNumber {
        options.smoothnessLevel = smoothnessLevel.floatValue
    }
    if let rednessLevel = map["rednessLevel"] as? NSNumber {
        options.rednessLevel = rednessLevel.floatValue
    }
    return options
}

func mapToAgoraImage(_ map: [String: Any]) -> AgoraImage {
    let image = AgoraImage()
    if let url = map["url"] as? String {
        if let url = URL(string: url) {
            image.url = url
        }
    }
    image.rect = mapToRect(map)
    return image
}

func mapToTranscodingUser(_ map: [String: Any]) -> AgoraLiveTranscodingUser {
    let user = AgoraLiveTranscodingUser()
    if let uid = map["uid"] as? NSNumber {
        user.uid = uid.uintValue
    }
    user.rect = mapToRect(map)
    if let zOrder = map["zOrder"] as? NSNumber {
        user.zOrder = zOrder.intValue
    }
    if let alpha = map["alpha"] as? NSNumber {
        user.alpha = alpha.doubleValue
    }
    if let audioChannel = map["audioChannel"] as? NSNumber {
        user.audioChannel = audioChannel.intValue
    }
    return user
}

func mapToColor(_ map: [String: Any]) -> UIColor {
    return UIColor(
        red: CGFloat((map["red"] as! NSNumber).intValue),
        green: CGFloat((map["green"] as! NSNumber).intValue),
        blue: CGFloat((map["blue"] as! NSNumber).intValue),
        alpha: 1.0
    )
}

func mapToLiveTranscoding(_ map: [String: Any]) -> AgoraLiveTranscoding {
    let transcoding = AgoraLiveTranscoding.default()
    transcoding.size = mapToSize(map)
    if let videoBitrate = map["videoBitrate"] as? NSNumber {
        transcoding.videoBitrate = videoBitrate.intValue
    }
    if let videoFramerate = map["videoFramerate"] as? NSNumber {
        transcoding.videoFramerate = videoFramerate.intValue
    }
    if let lowLatency = map["lowLatency"] as? Bool {
        transcoding.lowLatency = lowLatency
    }
    if let videoGop = map["videoGop"] as? NSNumber {
        transcoding.videoGop = videoGop.intValue
    }
    if let watermark = map["watermark"] as? [String: Any] {
        transcoding.watermark = mapToAgoraImage(watermark)
    }
    if let backgroundImage = map["backgroundImage"] as? [String: Any] {
        transcoding.backgroundImage = mapToAgoraImage(backgroundImage)
    }
    if let audioSampleRate = map["audioSampleRate"] as? NSNumber {
        if let audioSampleRate = AgoraAudioSampleRateType(rawValue: audioSampleRate.intValue) {
            transcoding.audioSampleRate = audioSampleRate
        }
    }
    if let audioBitrate = map["audioBitrate"] as? NSNumber {
        transcoding.audioBitrate = audioBitrate.intValue
    }
    if let audioChannels = map["audioChannels"] as? NSNumber {
        transcoding.audioChannels = audioChannels.intValue
    }
    if let audioCodecProfile = map["audioCodecProfile"] as? NSNumber {
        if let audioCodecProfile = AgoraAudioCodecProfileType(rawValue: audioCodecProfile.intValue) {
            transcoding.audioCodecProfile = audioCodecProfile
        }
    }
    if let videoCodecProfile = map["videoCodecProfile"] as? NSNumber {
        if let videoCodecProfile = AgoraVideoCodecProfileType(rawValue: videoCodecProfile.intValue) {
            transcoding.videoCodecProfile = videoCodecProfile
        }
    }
    if let backgroundColor = map["backgroundColor"] as? [String: Any] {
        transcoding.backgroundColor = mapToColor(backgroundColor)
    }
    if let userConfigExtraInfo = map["userConfigExtraInfo"] as? String {
        transcoding.transcodingExtraInfo = userConfigExtraInfo
    }
    if let transcodingUsers = map["transcodingUsers"] as? [Any] {
        transcodingUsers.forEach {
            if let item = $0 as? [String: Any] {
                transcoding.add(mapToTranscodingUser(item))
            }
        }
    }
    return transcoding
}

func mapToChannelMediaInfo(_ map: [String: Any]) -> AgoraChannelMediaRelayInfo {
    let info = AgoraChannelMediaRelayInfo()
    if let channelName = map["channelName"] as? String {
        info.channelName = channelName
    }
    if let token = map["token"] as? String {
        info.token = token
    }
    if let uid = map["uid"] as? NSNumber {
        info.uid = uid.uintValue
    }
    return info
}

func mapToChannelMediaRelayConfiguration(_ map: [String: Any]) -> AgoraChannelMediaRelayConfiguration {
    let config = AgoraChannelMediaRelayConfiguration()
    if let srcInfo = map["srcInfo"] as? [String: Any] {
        config.sourceInfo = mapToChannelMediaInfo(srcInfo)
    }
    if let destInfos = map["destInfos"] as? [Any] {
        destInfos.forEach {
            if let item = $0 as? [String: Any] {
                let info = mapToChannelMediaInfo(item)
                config.setDestinationInfo(info, forChannelName: info.channelName ?? "")
            }
        }
    }
    return config
}

func mapToLastmileProbeConfig(_ map: [String: Any]) -> AgoraLastmileProbeConfig {
    let config = AgoraLastmileProbeConfig()
    if let probeUplink = map["probeUplink"] as? Bool {
        config.probeUplink = probeUplink
    }
    if let probeDownlink = map["probeDownlink"] as? Bool {
        config.probeDownlink = probeDownlink
    }
    if let expectedUplinkBitrate = map["expectedUplinkBitrate"] as? NSNumber {
        config.expectedUplinkBitrate = expectedUplinkBitrate.uintValue
    }
    if let expectedDownlinkBitrate = map["expectedDownlinkBitrate"] as? NSNumber {
        config.expectedDownlinkBitrate = expectedDownlinkBitrate.uintValue
    }
    return config
}

func mapToWatermarkOptions(_ map: [String: Any]) -> WatermarkOptions {
    let options = WatermarkOptions()
    if let visibleInPreview = map["visibleInPreview"] as? Bool {
        options.visibleInPreview = visibleInPreview
    }
    if let positionInLandscapeMode = map["positionInLandscapeMode"] as? [String: Any] {
        options.positionInLandscapeMode = mapToRect(positionInLandscapeMode)
    }
    if let positionInPortraitMode = map["positionInPortraitMode"] as? [String: Any] {
        options.positionInPortraitMode = mapToRect(positionInPortraitMode)
    }
    return options
}

func mapToLiveInjectStreamConfig(_ map: [String: Any]) -> AgoraLiveInjectStreamConfig {
    let config = AgoraLiveInjectStreamConfig.default()
    config.size = mapToSize(map)
    if let videoGop = map["videoGop"] as? NSNumber {
        config.videoGop = videoGop.intValue
    }
    if let videoFramerate = map["videoFramerate"] as? NSNumber {
        config.videoFramerate = videoFramerate.intValue
    }
    if let videoBitrate = map["videoBitrate"] as? NSNumber {
        config.videoBitrate = videoBitrate.intValue
    }
    if let audioSampleRate = map["audioSampleRate"] as? NSNumber {
        if let audioSampleRate = AgoraAudioSampleRateType(rawValue: audioSampleRate.intValue) {
            config.audioSampleRate = audioSampleRate
        }
    }
    if let audioBitrate = map["audioBitrate"] as? NSNumber {
        config.audioBitrate = audioBitrate.intValue
    }
    if let audioChannels = map["audioChannels"] as? NSNumber {
        config.audioChannels = audioChannels.intValue
    }
    return config
}

func mapToCameraCapturerConfiguration(_ map: [String: Any]) -> AgoraCameraCapturerConfiguration {
    let config = AgoraCameraCapturerConfiguration()
    if let preference = map["preference"] as? NSNumber {
        if let preference = AgoraCameraCaptureOutputPreference(rawValue: preference.intValue) {
            config.preference = preference
        }
    }
    if let captureWidth = map["captureWidth"] as? NSNumber {
        config.captureWidth = captureWidth.int32Value
    }
    if let captureHeight = map["captureHeight"] as? NSNumber {
        config.captureHeight = captureHeight.int32Value
    }
    if let cameraDirection = map["cameraDirection"] as? NSNumber {
        if let cameraDirection = AgoraCameraDirection(rawValue: cameraDirection.intValue) {
            config.cameraDirection = cameraDirection
        }
    }
    return config
}

func mapToChannelMediaOptions(_ map: [String: Any]) -> AgoraRtcChannelMediaOptions {
    let options = AgoraRtcChannelMediaOptions()
    if let autoSubscribeAudio = map["autoSubscribeAudio"] as? Bool {
        options.autoSubscribeAudio = autoSubscribeAudio
    }
    if let autoSubscribeVideo = map["autoSubscribeVideo"] as? Bool {
        options.autoSubscribeVideo = autoSubscribeVideo
    }
    if let publishLocalAudio = map["publishLocalAudio"] as? Bool {
        options.publishLocalAudio = publishLocalAudio
    }
    if let publishLocalVideo = map["publishLocalVideo"] as? Bool {
        options.publishLocalVideo = publishLocalVideo
    }
    return options
}

func mapToRtcEngineConfig(_ map: [String: Any]) -> AgoraRtcEngineConfig {
    let config = AgoraRtcEngineConfig()
    config.appId = map["appId"] as? String
    if let list = map["areaCode"] as? [Any] {
        var areaCode: UInt = 0
        for i in list {
            if let code = i as? NSNumber {
                areaCode |= code.uintValue
            }
        }
        config.areaCode = areaCode
    }
    if let logConfig = map["logConfig"] as? [String: Any] {
        config.logConfig = mapToLogConfig(logConfig)
    }
    return config
}

func mapToAudioRecordingConfiguration(_ map: [String: Any]) -> AgoraAudioRecordingConfiguration {
    let config = AgoraAudioRecordingConfiguration()
    if let filePath = map["filePath"] as? String {
        config.filePath = filePath
    }
    if let recordingQuality = map["recordingQuality"] as? NSNumber {
        if let recordingQuality = AgoraAudioRecordingQuality(rawValue: recordingQuality.intValue) {
            config.recordingQuality = recordingQuality
        }
    }
    if let recordingPosition = map["recordingPosition"] as? NSNumber {
        if let recordingPosition = AgoraAudioRecordingPosition(rawValue: recordingPosition.intValue) {
            config.recordingPosition = recordingPosition
        }
    }
    if let recordingSampleRate = map["recordingSampleRate"] as? NSNumber {
        config.recordingSampleRate = recordingSampleRate.intValue
    }
    return config
}

func mapToEncryptionConfig(_ map: [String: Any]) -> AgoraEncryptionConfig {
    let config = AgoraEncryptionConfig()
    if let encryptionMode = map["encryptionMode"] as? NSNumber {
        if let encryptionMode = AgoraEncryptionMode(rawValue: encryptionMode.intValue) {
            config.encryptionMode = encryptionMode
        }
    }
    if let encryptionKey = map["encryptionKey"] as? String {
        config.encryptionKey = encryptionKey
    }
    if let list = map["encryptionKdfSalt"] as? [Any] {
        var encryptionKdfSalt: [UInt8] = []
        for i in list.indices {
            if let item = list[i] as? NSNumber {
                encryptionKdfSalt[i] = item.uint8Value
            }
        }
        config.encryptionKdfSalt = Data(bytes: encryptionKdfSalt)
    }
    return config
}

func mapToRhythmPlayerConfig(_ map: [String: Any]) -> AgoraRtcRhythmPlayerConfig {
    let config = AgoraRtcRhythmPlayerConfig()
    if let beatsPerMeasure = map["beatsPerMeasure"] as? NSNumber {
        config.beatsPerMeasure = beatsPerMeasure.uintValue
    }
    if let beatsPerMinute = map["beatsPerMinute"] as? NSNumber {
        config.beatsPerMinute = beatsPerMinute.uintValue
    }
    if let publish = map["publish"] as? NSNumber {
        config.publish = publish.boolValue
    }
    return config
}

func mapToClientRoleOptions(_ map: [String: Any]) -> AgoraClientRoleOptions {
    let options = AgoraClientRoleOptions()
    if let audienceLatencyLevel = map["audienceLatencyLevel"] as? NSNumber {
        if let audienceLatencyLevel = AgoraAudienceLatencyLevelType(rawValue: audienceLatencyLevel.intValue) {
            options.audienceLatencyLevel = audienceLatencyLevel
        }
    }
    return options
}

func mapToLogConfig(_ map: [String: Any]) -> AgoraLogConfig {
    let config = AgoraLogConfig()
    config.filePath = map["filePath"] as? String
    if let fileSize = map["fileSize"] as? NSNumber {
        config.fileSize = fileSize.intValue
    }
    if let level = map["level"] as? NSNumber {
        if let level = AgoraLogLevel(rawValue: level.intValue) {
            config.level = level
        }
    }
    return config
}

func mapToDataStreamConfig(_ map: [String: Any]) -> AgoraDataStreamConfig {
    let config = AgoraDataStreamConfig()
    if let syncWithAudio = map["syncWithAudio"] as? Bool {
        config.syncWithAudio = syncWithAudio
    }
    if let ordered = map["ordered"] as? Bool {
        config.ordered = ordered
    }
    return config
}

File: BeanCovertor.kt
package io.agora.rtc.base

import io.agora.rtc.RtcEngineConfig
import io.agora.rtc.audio.AgoraRhythmPlayerConfig
import io.agora.rtc.audio.AudioRecordingConfiguration
import io.agora.rtc.internal.EncryptionConfig
import io.agora.rtc.internal.LastmileProbeConfig
import io.agora.rtc.live.LiveInjectStreamConfig
import io.agora.rtc.live.LiveTranscoding
import io.agora.rtc.live.LiveTranscoding.TranscodingUser
import io.agora.rtc.models.ChannelMediaOptions
import io.agora.rtc.models.ClientRoleOptions
import io.agora.rtc.models.DataStreamConfig
import io.agora.rtc.models.EchoTestConfiguration
import io.agora.rtc.video.*

fun mapToVideoDimensions(map: Map<*, *>): VideoEncoderConfiguration.VideoDimensions {
  return VideoEncoderConfiguration.VideoDimensions().apply {
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
  }
}

fun mapToVideoEncoderConfiguration(map: Map<*, *>): VideoEncoderConfiguration {
  return VideoEncoderConfiguration().apply {
    (map["dimensions"] as? Map<*, *>)?.let { dimensions = mapToVideoDimensions(it) }
    (map["frameRate"] as? Number)?.let { frameRate = it.toInt() }
    (map["minFrameRate"] as? Number)?.let { minFrameRate = it.toInt() }
    (map["bitrate"] as? Number)?.let { bitrate = it.toInt() }
    (map["minBitrate"] as? Number)?.let { minBitrate = it.toInt() }
    (map["orientationMode"] as? Number)?.let { orientationMode = intToOrientationMode(it.toInt()) }
    (map["degradationPrefer"] as? Number)?.let {
      degradationPrefer = intToDegradationPreference(it.toInt())
    }
    (map["mirrorMode"] as? Number)?.let { mirrorMode = it.toInt() }
  }
}

fun mapToBeautyOptions(map: Map<*, *>): BeautyOptions {
  return BeautyOptions().apply {
    (map["lighteningContrastLevel"] as? Number)?.let { lighteningContrastLevel = it.toInt() }
    (map["lighteningLevel"] as? Number)?.let { lighteningLevel = it.toFloat() }
    (map["smoothnessLevel"] as? Number)?.let { smoothnessLevel = it.toFloat() }
    (map["rednessLevel"] as? Number)?.let { rednessLevel = it.toFloat() }
  }
}

fun mapToAgoraImage(map: Map<*, *>): AgoraImage {
  return AgoraImage().apply {
    (map["url"] as? String)?.let { url = it }
    (map["x"] as? Number)?.let { x = it.toInt() }
    (map["y"] as? Number)?.let { y = it.toInt() }
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
  }
}

fun mapToTranscodingUser(map: Map<*, *>): TranscodingUser {
  return TranscodingUser().apply {
    (map["uid"] as? Number)?.let { uid = it.toNativeUInt() }
    (map["x"] as? Number)?.let { x = it.toInt() }
    (map["y"] as? Number)?.let { y = it.toInt() }
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
    (map["zOrder"] as? Number)?.let { zOrder = it.toInt() }
    (map["alpha"] as? Number)?.let { alpha = it.toFloat() }
    (map["audioChannel"] as? Number)?.let { audioChannel = it.toInt() }
  }
}

fun mapToColor(map: Map<*, *>): Int {
  return ((map["red"] as Number).toInt() shl 16) + ((map["green"] as Number).toInt() shl 8) + (map["blue"] as Number).toInt()
}

fun mapToLiveTranscoding(map: Map<*, *>): LiveTranscoding {
  return LiveTranscoding().apply {
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
    (map["videoBitrate"] as? Number)?.let { videoBitrate = it.toInt() }
    (map["videoFramerate"] as? Number)?.let { videoFramerate = it.toInt() }
    (map["lowLatency"] as? Boolean)?.let { lowLatency = it }
    (map["videoGop"] as? Number)?.let { videoGop = it.toInt() }
    (map["watermark"] as? Map<*, *>)?.let { watermark = mapToAgoraImage(it) }
    (map["backgroundImage"] as? Map<*, *>)?.let { backgroundImage = mapToAgoraImage(it) }
    (map["audioSampleRate"] as? Number)?.let {
      audioSampleRate = intToLiveTranscodingAudioSampleRate(it.toInt())
    }
    (map["audioBitrate"] as? Number)?.let { audioBitrate = it.toInt() }
    (map["audioChannels"] as? Number)?.let { audioChannels = it.toInt() }
    (map["audioCodecProfile"] as? Number)?.let {
      audioCodecProfile = intToAudioCodecProfile(it.toInt())
    }
    (map["videoCodecProfile"] as? Number)?.let {
      videoCodecProfile = intToVideoCodecProfile(it.toInt())
    }
    (map["backgroundColor"] as? Map<*, *>)?.let { backgroundColor = mapToColor(it) }
    (map["userConfigExtraInfo"] as? String)?.let { userConfigExtraInfo = it }
    (map["transcodingUsers"] as? List<*>)?.let { list ->
      list.forEach { item ->
        (item as? Map<*, *>)?.let {
          addUser(mapToTranscodingUser(it))
        }
      }
    }
  }
}

fun mapToChannelMediaInfo(map: Map<*, *>): ChannelMediaInfo {
  return ChannelMediaInfo(
    map["channelName"] as? String,
    map["token"] as? String,
    (map["uid"] as Number).toNativeUInt()
  )
}

fun mapToChannelMediaRelayConfiguration(map: Map<*, *>): ChannelMediaRelayConfiguration {
  return ChannelMediaRelayConfiguration().apply {
    (map["srcInfo"] as? Map<*, *>)?.let { setSrcChannelInfo(mapToChannelMediaInfo(it)) }
    (map["destInfos"] as? List<*>)?.let { list ->
      list.forEach { item ->
        (item as? Map<*, *>)?.let {
          val info = mapToChannelMediaInfo(it)
          setDestChannelInfo(info.channelName, info)
        }
      }
    }
  }
}

fun mapToLastmileProbeConfig(map: Map<*, *>): LastmileProbeConfig {
  return LastmileProbeConfig().apply {
    (map["probeUplink"] as? Boolean)?.let { probeUplink = it }
    (map["probeDownlink"] as? Boolean)?.let { probeDownlink = it }
    (map["expectedUplinkBitrate"] as? Number)?.let { expectedUplinkBitrate = it.toInt() }
    (map["expectedDownlinkBitrate"] as? Number)?.let { expectedUplinkBitrate = it.toInt() }
  }
}

fun mapToRectangle(map: Map<*, *>): WatermarkOptions.Rectangle {
  return WatermarkOptions.Rectangle().apply {
    (map["x"] as? Number)?.let { x = it.toInt() }
    (map["y"] as? Number)?.let { y = it.toInt() }
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
  }
}

fun mapToWatermarkOptions(map: Map<*, *>): WatermarkOptions {
  return WatermarkOptions().apply {
    (map["visibleInPreview"] as? Boolean)?.let { visibleInPreview = it }
    (map["positionInLandscapeMode"] as? Map<*, *>)?.let {
      positionInLandscapeMode = mapToRectangle(it)
    }
    (map["positionInPortraitMode"] as? Map<*, *>)?.let {
      positionInPortraitMode = mapToRectangle(it)
    }
  }
}

fun mapToLiveInjectStreamConfig(map: Map<*, *>): LiveInjectStreamConfig {
  return LiveInjectStreamConfig().apply {
    (map["width"] as? Number)?.let { width = it.toInt() }
    (map["height"] as? Number)?.let { height = it.toInt() }
    (map["videoGop"] as? Number)?.let { videoGop = it.toInt() }
    (map["videoFramerate"] as? Number)?.let { videoFramerate = it.toInt() }
    (map["videoBitrate"] as? Number)?.let { videoBitrate = it.toInt() }
    (map["audioSampleRate"] as? Number)?.let {
      audioSampleRate = intToLiveInjectStreamConfigAudioSampleRate(it.toInt())
    }
    (map["audioBitrate"] as? Number)?.let { audioBitrate = it.toInt() }
    (map["audioChannels"] as? Number)?.let { audioChannels = it.toInt() }
  }
}

fun mapToRhythmPlayerConfig(map: Map<*, *>): AgoraRhythmPlayerConfig {
  return AgoraRhythmPlayerConfig().apply {
    (map["beatsPerMeasure"] as? Number)?.let { beatsPerMeasure = it.toInt() }
    (map["beatsPerMinute"] as? Number)?.let { beatsPerMinute = it.toInt() }
    (map["publish"] as? Boolean)?.let { publish = it }
  }
}

fun mapToCameraCapturerConfiguration(map: Map<*, *>): CameraCapturerConfiguration {
  return CameraCapturerConfiguration(
    intToCapturerOutputPreference((map["preference"] as Number).toInt()),
    intToCameraDirection((map["cameraDirection"] as Number).toInt())
  ).apply {
    dimensions = CameraCapturerConfiguration.CaptureDimensions()
    (map["captureWidth"] as? Number)?.toInt()?.let { dimensions.width = it }
    (map["captureHeight"] as? Number)?.toInt()?.let { dimensions.height = it }
  }
}

fun mapToChannelMediaOptions(map: Map<*, *>): ChannelMediaOptions {
  return ChannelMediaOptions().apply {
    (map["autoSubscribeAudio"] as? Boolean)?.let { autoSubscribeAudio = it }
    (map["autoSubscribeVideo"] as? Boolean)?.let { autoSubscribeVideo = it }
    (map["publishLocalAudio"] as? Boolean)?.let { publishLocalAudio = it }
    (map["publishLocalVideo"] as? Boolean)?.let { publishLocalVideo = it }
  }
}

fun mapToRtcEngineConfig(map: Map<*, *>): RtcEngineConfig {
  return RtcEngineConfig().apply {
    mAppId = map["appId"] as String
    (map["areaCode"] as? Number)?.toInt()?.let { mAreaCode = it }
    (map["logConfig"] as? Map<*, *>)?.let { mLogConfig = mapToLogConfig(it) }
  }
}

fun mapToAudioRecordingConfiguration(map: Map<*, *>): AudioRecordingConfiguration {
  return AudioRecordingConfiguration().apply {
    (map["filePath"] as? String)?.let { filePath = it }
    (map["recordingQuality"] as? Number)?.let { recordingQuality = it.toInt() }
    (map["recordingPosition"] as? Number)?.let { recordingPosition = it.toInt() }
    (map["recordingSampleRate"] as? Number)?.let { recordingSampleRate = it.toInt() }
  }
}

fun mapToEncryptionConfig(map: Map<*, *>): EncryptionConfig {
  return EncryptionConfig().apply {
    (map["encryptionMode"] as? Number)?.let { encryptionMode = intToEncryptionMode(it.toInt()) }
    (map["encryptionKey"] as? String)?.let { encryptionKey = it }
    (map["encryptionKdfSalt"] as? List<*>)?.let { list ->
      for (i in list.indices) {
        (list[i] as? Number)?.let {
          encryptionKdfSalt[i] = it.toByte()
        }
      }
    }
  }
}

fun mapToClientRoleOptions(map: Map<*, *>): ClientRoleOptions {
  return ClientRoleOptions().apply {
    (map["audienceLatencyLevel"] as? Number)?.let { audienceLatencyLevel = it.toInt() }
  }
}

fun mapToLogConfig(map: Map<*, *>): RtcEngineConfig.LogConfig {
  return RtcEngineConfig.LogConfig().apply {
    (map["filePath"] as? String)?.let { filePath = it }
    (map["fileSize"] as? Number)?.let { fileSize = it.toInt() }
    (map["level"] as? Number)?.let { level = it.toInt() }
  }
}

fun mapToDataStreamConfig(map: Map<*, *>): DataStreamConfig {
  return DataStreamConfig().apply {
    (map["syncWithAudio"] as? Boolean)?.let { syncWithAudio = it }
    (map["ordered"] as? Boolean)?.let { ordered = it }
  }
}

fun mapToVirtualBackgroundSource(map: Map<*, *>): VirtualBackgroundSource {
  return VirtualBackgroundSource().apply {
    (map["backgroundSourceType"] as? Number)?.let { backgroundSourceType = it.toInt() }
    (map["color"] as? Map<*, *>)?.let { color = mapToColor(it) }
    (map["source"] as? String)?.let { source = it }
    (map["blur_degree"] as? Int)?.let { blur_degree = it }
  }
}

fun mapToEchoTestConfiguration(map: Map<*, *>): EchoTestConfiguration {
  return EchoTestConfiguration().apply {
    (map["enableAudio"] as? Boolean)?.let { enableAudio = it }
    (map["enableVideo"] as? Boolean)?.let { enableVideo = it }
    (map["token"] as? String)?.let { token = it }
    (map["channelId"] as? String)?.let { channelId = it }
  }
}
